homeassistant/components/device_tracker/automatic.py

```diff
@@ -38,65 +44,183 @@
 
+AUTOMATIC_CONFIG_FILE = 'automatic/session-{}.json'
+
+DATA_CONFIGURING = 'automatic_configurator_clients'
+DATA_REFRESH_TOKEN = 'refresh_token'
+
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
     vol.Required(CONF_CLIENT_ID): cv.string,
     vol.Required(CONF_SECRET): cv.string,
-    vol.Required(CONF_USERNAME): cv.string,
-    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Inclusive(CONF_USERNAME, 'auth'): cv.string,
+    vol.Inclusive(CONF_PASSWORD, 'auth'): cv.string,
+    vol.Optional(CONF_CURRENT_LOCATION, default=False): cv.boolean,
     vol.Optional(CONF_DEVICES, default=None): vol.All(
         cv.ensure_list, [cv.string])
 })
 
 
+def _get_refresh_token_from_file(hass, filename):
+    """Attempt to load session data from file."""
+    path = hass.config.path(filename)
+
+    if not os.path.isfile(path):
+        return None
+
+    try:
+        with open(path) as data_file:
+            data = json.load(data_file)
+            if data is None:
+                return None
+
+            return data.get(DATA_REFRESH_TOKEN)
+    except ValueError:
+        return None
+
+
+def _write_refresh_token_to_file(hass, filename, refresh_token):
+    """Attempt to store session data to file."""
+    path = hass.config.path(filename)
+
+    os.makedirs(os.path.dirname(path), exist_ok=True)
+    with open(path, 'w+') as data_file:
+        json.dump({
+            DATA_REFRESH_TOKEN: refresh_token
+        }, data_file)
+
+
 @asyncio.coroutine
 def async_setup_scanner(hass, config, async_see, discovery_info=None):
     """Validate the configuration and return an Automatic scanner."""
     import aioautomatic
 
+    hass.http.register_view(AutomaticAuthCallbackView())
+
+    scope = FULL_SCOPE if config.get(CONF_CURRENT_LOCATION) else DEFAULT_SCOPE
+
     client = aioautomatic.Client(
         client_id=config[CONF_CLIENT_ID],
         client_secret=config[CONF_SECRET],
         client_session=async_get_clientsession(hass),
         request_kwargs={'timeout': DEFAULT_TIMEOUT})
-    try:
-        try:
-            session = yield from client.create_session_from_password(
-                FULL_SCOPE, config[CONF_USERNAME], config[CONF_PASSWORD])
-        except aioautomatic.exceptions.ForbiddenError as exc:
-            if not str(exc).startswith("invalid_scope"):
-                raise exc
-            _LOGGER.info("Client not authorized for current_location scope. "
-                         "location:updated events will not be received.")
-            session = yield from client.create_session_from_password(
-                DEFAULT_SCOPE, config[CONF_USERNAME], config[CONF_PASSWORD])
 
+    filename = AUTOMATIC_CONFIG_FILE.format(config[CONF_CLIENT_ID])
+    refresh_token = yield from hass.async_add_job(
+        _get_refresh_token_from_file, hass, filename)
+
+    @asyncio.coroutine
+    def initialize_data(session):
+        """Initialize the AutomaticData object from the created session."""
+        hass.async_add_job(
```

Comment:
Yeah, when we open a new session with the old refresh token, we get a new refresh token as part of the response.